[jit] refactor self to be a class again

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22504 [WIP] classtypes hold weak_ptr
* #22692 [jit] Make traced fns also go into the global python CU
* #22221 [jit] _script_compile and _script_class_compile add to the python CU
* #22667 [jit] make CompilationUnit::define return defined functions
* **#22207 [jit] refactor self to be a class again**
* #22206 [jit] Give functions qualified names

I want to get the `type` of self (now that everything is ultimately a
class type).

Differential Revision: [D15998758](https://our.internmc.facebook.com/intern/diff/D15998758)

suo committed:
[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

zdevito commented on torch/csrc/jit/script/compilation_unit.h:
```diff
@@ -30,7 +30,11 @@ struct SugaredValue;
+struct Self {
+  virtual ~Self() {}
+  virtual std::shared_ptr<SugaredValue> makeSugared(Value* v) const = 0;
+  virtual ClassTypePtr getClassType() const = 0;
```
This is unused in the PR, which makes it hard for me to evaluate this PR on its own. I don't like having to do this because it re-introduces an annoying complexity for this self argument. I don't understand why we can't extract the type for the value (v->type()) or from the schema.

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head

suo commented:
Re-requesting review. I mentioned on the above PR: we need to be able to register the fact that the method exists before compiling it, so that methods can find each other.

zdevito reviewed (approved):
I think it would be clearer as a separate method, but my opinion here is not very strong.

suo committed:
Update on "[jit] refactor self to be a class again"

[jit] refactor self to be a class again

I want to get the `type` of self (now that everything is ultimately a
class type).

gh-metadata: pytorch pytorch 22207 gh/suo/71/head
